docs/architecture/adr-032-typed-events.md

```diff
@@ -28,218 +28,106 @@ The end of this proposal contains a detailed example of how to consume events af
-    ABCIEvent() Event
-}
-
-// BaseModuleEvent contains information about the Module and Action of an event
-type BaseModuleEvent struct {
-    Module string
-    Action string
-}
-```
-
-The `BaseModuleEvent` struct will be used for basic context about the event. It aids in routing events to their 
-`Module` and `Action` specific event parsers.
-
-__Step 2__:  Implement additional functionality in the `types` package: utility functions and a parser to route the event to its proper module.
-
-When we subscribe to emitted events on the tendermint websocket, they are emitted in the form of an `abci.Event`. The parser will process this event using `sdk.NewSDKEvent(abci.Event)` to enable passing of the processed event to the proper module.
-
-```go
-// types/events.go
-
-// SDKEvent contains the string representation of the event and the module information
-type SDKEvent struct {
-    Sev  StringEvent
-    Base BaseModuleEvent
-}
-
-// NewSDKEvent parses abci.Event into an sdk.SDKEvent
-func NewSDKEvent(bev abci.Event) (ev SDKEvent, err error) {
-    sev := StringifyEvent(bev)
-    module, err := GetEventString(sev.Attributes, AttributeKeyModule)
-    if err != nil {
-        return
-    }
-    action, err := GetEventString(sev.Attributes, AttributeKeyAction)
-    if err != nil {
-        return
-    }
-    return SDKEvent{sev, BaseModuleEvent{module, action}}, nil
-}
+// EmitTypedEvent takes typed event and emits converting it into sdk.Event 
+func (em *EventManager) EmitTypedEvent(event proto.Message) error {
+	evtType := proto.MessageName(event)
+	evtJSON, err := codec.ProtoMarshalJSON(event)
+	if err != nil {
+		return err
+	}
+
+	var attrMap map[string]json.RawMessage
+	err = json.Unmarshal(evtJSON, &attrMap)
+	if err != nil {
+		return err
+	}
+
+	var attrs []abci.EventAttribute
+	for k, v := range attrMap {
+		attrs = append(attrs, abci.EventAttribute{
+			Key:   []byte(k),
+			Value: v,
+		})
+	}
+
+	em.EmitEvent(Event{
+		Type:       evtType,
+		Attributes: attrs,
+	})
 
-// GetEventString take sdk attributes, key and returns value for that key. 
-func GetEventString(attrs []Attribute, key string) (string, error) {
-    for _, attr := range attrs {
-        if attr.Key == key {
-            return attr.Value, nil
-        }
-    }
-    return "", fmt.Errorf("not found")
+	return nil
 }
 
-// GetEventUint64 take sdk attributes, key and returns uint64 value. 
-// Returns error incase of failure.
-func GetEventUint64(attrs []Attribute, key string) (uint64, error) {
-    sval, err := GetEventString(attrs, key)
-    if err != nil {
-        return 0, err
-    }
-    return strconv.ParseUint(sval, 10, 64)
+// ParseTypedEvent converts abci.Event back to typed event
+func ParseTypedEvent(event abci.Event) (proto.Message, error) {
+	concreteGoType := proto.MessageType(event.Type)
+	if concreteGoType == nil {
+		return nil, fmt.Errorf("failed to retrieve the message of type %q", event.Type)
+	}
+
+	value := reflect.New(concreteGoType).Elem()
```

Comment:
If concreteType is a pointer, then New creates a nil, and we cannot .Elem() of a nil.

What we should do is:
- if concreteType is a pointer, then do `reflect.New(concreteType.Elem())`
- if not, then `reflect.Zero(concreteType)`

And then it should work.
```suggestion
 if concreteGoType.Kind() == reflect.Ptr {
   value = reflect.New(concreteType.Elem());
 } else {
   value = reflect.Zero(concreteType)
 }
```

Reply:
Sure. Thanks @amaurymartiny